python/tvm/autotvm/tuner/ga_tuner.py

```diff
@@ -0,0 +1,125 @@
+                tmp_gene = []
+                for j in range(self.n_subspace):
+                    tmp_gene.append(np.random.randint(self.dim_subspaces[j]))
+            self.genes.append(tmp_gene)
+            self.visited.add(self._gene2index(tmp_gene))
+
+    def next_batch(self, batch_size):
+        ret = []
+        for _ in range(batch_size):
+            gene = self.genes[self.trial_pt % self.pop_size]
+            self.trial_pt += 1
+            ret.append(self.space.get(self._gene2index(gene)))
+
+        return ret
+
+    def update(self, inputs, results):
+        for inp, res in zip(inputs, results):
+            if res.error_no == 0:
+                y = inp.task.flop / np.mean(res.costs)
+                self.scores.append(y)
+            else:
+                self.scores.append(0)
+
+        if len(self.scores) >= len(self.genes):
+            genes = self.genes + self.elites
+            scores = np.array(self.scores[:len(self.genes)] + self.elite_scores)
+            assert len(genes) == len(scores), "%d vs %d" % (len(genes), len(scores))
+            indices = np.arange(len(genes))
+
+            # reserve elite
+            self.elites, self.elite_scores = [], []
+            elites = np.argpartition(scores, -self.elite_num)[-self.elite_num:]
+            for ind in elites:
+                self.elites.append(genes[ind])
+                self.elite_scores.append(scores[ind])
+
+            # cross over
+            scores /= np.max(scores)
+            probs = scores / np.sum(scores)
+            tmp_genes = []
+            for i in range(self.pop_size):
+                p1, p2 = np.random.choice(indices, size=2, replace=False, p=probs)
+                p1, p2 = genes[p1], genes[p2]
+                point = np.random.randint(self.n_subspace)
+                tmp_gene = p1[:point] + p2[point:]
+                tmp_genes.append(tmp_gene)
+
+            # mutation
+            next_genes = []
+            for i in range(self.pop_size):
+                for j in range(self.n_subspace):
+                    if np.random.random() < self.mutation_prob:
+                        tmp_genes[i][j] = np.random.randint(self.dim_subspaces[j])
+
+                if len(self.visited) < len(self.space):
+                    while self._gene2index(tmp_genes[i]) in self.visited:
+                        j = np.random.randint(self.n_subspace)
+                        tmp_genes[i][j] = np.random.randint(self.dim_subspaces[j])
+                    next_genes.append(tmp_genes[i])
+                    self.visited.add(self._gene2index(tmp_genes[i]))
+                else:
+                    break
+
+            self.genes = next_genes
+            self.trial_pt = 0
+
+    def has_next(self):
+        return len(self.visited) - (len(self.genes) - self.trial_pt) < len(self.space)
+
+    def _gene2index(self, gene):
+        ind = 0
+        for j in range(self.n_subspace):
+            ind = (ind + gene[j]) * self.dim_subspaces[j+1]
```

Comment:
If we would remove the tail `[1]` from self.n_subspace, this part would be something like as follows.
```
    for g, d in zip(gene, self.dim_subspaces[1:] + [1]):
        ind = (ind + g) * d
```